Fix an issue where functional collectives don't force fx stride on inputs when compiled

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146467

Fixes https://github.com/pytorch/pytorch/issues/146416

Also added contiguity checks in the C++ functional collective ops to prevent striding issues introduced during compilation manifest as silent correctness issues.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

Commit:
Fix an issue where functional collectives don't force fx stride on inputs when compiled

[ghstack-poisoned]

Commit:
Update on "Fix an issue where functional collectives don't force fx stride on inputs when compiled"


Fixes https://github.com/pytorch/pytorch/issues/146416

Also added contiguity checks in the C++ functional collective ops to prevent striding issues introduced during compilation manifest as silent correctness issues.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

lw reviewed (approved):
I'm not sure I understand all the implications of this change, but thanks for the fix!

lw commented:
Should this PR be tagged for inclusion in v2.6.1?

shunting314 commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -493,6 +495,62 @@ def join(self):
+            in_, w, out_grad
+        )
+
+        assert torch.allclose(compile_w_grad, eager_w_grad)
```
nit: self.assertTrue(...)

yifuwang commented:
@pytorchbot rebase

Commit:
Update

[ghstack-poisoned]

Commit:
Update on "Fix an issue where functional collectives don't force fx stride on inputs when compiled"


Fixes https://github.com/pytorch/pytorch/issues/146416

Also added contiguity checks in the C++ functional collective ops to prevent striding issues introduced during compilation manifest as silent correctness issues.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

Commit:
Update on "Fix an issue where functional collectives don't force fx stride on inputs when compiled"


Fixes https://github.com/pytorch/pytorch/issues/146416

Also added contiguity checks in the C++ functional collective ops to prevent striding issues introduced during compilation manifest as silent correctness issues.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

yifuwang commented:
@pytorchbot merge

yifuwang commented:
@pytorchbot merge
